lang/cs/Org.Apache.REEF.IO/DataCache.cs

```diff
@@ -0,0 +1,479 @@
+using Org.Apache.REEF.IO.Files;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Utilities;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IO
+{
+    /// <summary>
+    /// A class that manages the caching of data at different layers of storage medium.
+    /// </summary>
+    [Unstable("0.15", "API contract may change.")]
+    public sealed class DataCache<T>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(DataCache<>));
+
+        private readonly IInputDataMover<T> _inputDataMover;
+
+        private Optional<IDirectoryInfo> _diskDirectory = Optional<IDirectoryInfo>.Empty();
+        private Optional<IReadOnlyCollection<MemoryStream>> _memStreams = Optional<IReadOnlyCollection<MemoryStream>>.Empty();
+        private Optional<IReadOnlyCollection<T>> _materialized = Optional<IReadOnlyCollection<T>>.Empty();
+
+        private CacheLevel _lowestCacheLevel = CacheLevel.NotLocal;
+
+        [Inject]
+        private DataCache(IInputDataMover<T> inputDataMover)
+        {
+            _inputDataMover = inputDataMover;
+        }
+
+        /// <summary>
+        /// The current level at which the data is cached.
+        /// </summary>
+        public CacheLevel CacheLevel
+        {
+            get { return _lowestCacheLevel; }
+        }
+
+        /// <summary>
+        /// The directory where cached data resides, if the data is cached on disk.
+        /// </summary>
+        internal Optional<IDirectoryInfo> DiskDirectory
+        {
+            get { return _diskDirectory; }
+        }
+
+        /// <summary>
+        /// The collection of memory streams where cached data resides, if the data is
+        /// cached in memory.
+        /// </summary>
+        internal Optional<IReadOnlyCollection<MemoryStream>> MemoryStreams
+        {
+            get { return _memStreams; }
+        }
+
+        /// <summary>
+        /// The cached collection of "materialized" objects, if the data is cached
+        /// in memory and deserialized.
+        /// </summary>
+        internal Optional<IReadOnlyCollection<T>> Materialized
+        {
+            get { return _materialized; }
+        }
+
+        /// <summary>
+        /// Clears all levels of cache.
+        /// </summary>
+        public void Clear()
+        {
+            while (CleanCacheAtLevel(CacheLevel) != CacheLevel.NotLocal)
+            {
+            }
+        }
+
+        /// <summary>
+        /// Cache the data to a <see cref="CacheLevel"/>.
+        /// If the data is already cached at a lower level, no action will be taken.
+        /// </summary>
+        /// <param name="cacheToLevel">Level to cache to.</param>
+        /// <param name="shouldCleanHigherLevelCache">Whether or not to clean the higher level caches.</param>
+        /// <returns>The cached level.</returns>
```

Comment:
It will return the lowest cache level if `_lowestCacheLevel` is already lower than or equal to than `cacheToLevel`.